src/operator/numpy/np_indexing_op.cu

```diff
@@ -0,0 +1,574 @@
+      j = j % K;
+      j += (j < 0) ? K : 0;
+
+      for (int64_t k = 0; k < M; k++){
+        out_data[i * M + k] = in_data[j * M + k];
+      }
+    }
+};
+
+struct AdvancedIndexingTakeMultiDimensionGPU {
+    // assume that idx have been flattened to a 1-D tensor (N,)
+    // assume that out_data and in_data have been flattened to 2-D tensors, (N, M) and (K, M)
+    // M is the number of columns of in_data and out_data
+    // K is the number of rows of in_data
+    // i is the index of out_data
+    template<typename DType, typename IType>
+    MSHADOW_XINLINE static void Map(int i, DType* out_data, const DType* in_data,
+                                    const IType* idx, const int64_t M, const int64_t K) {
+      int64_t j = static_cast<int64_t>(idx[i]);
+      j = j % K;
+      j += (j < 0) ? K : 0;
+
+      for (int64_t k = 0; k < M; k++){
+        out_data[i * M + k] = in_data[(i * k + j) * M + k];
+      }
+    }
+};
+
+template<>
+inline void AdvancedIndexingOpForward<gpu>(const nnvm::NodeAttrs& attrs,
+                                    const OpContext &ctx,
+                                    const std::vector<NDArray> &inputs,
+                                    const std::vector<OpReqType> &req,
+                                    const std::vector<NDArray> &outputs) {
+  using namespace mshadow;
+  CHECK_EQ(inputs.size(), 2U);
+  CHECK_EQ(outputs.size(), 1U);
+
+  if (inputs[np_indexing_::kIdx].dtype() == mshadow::kBool) {
+    CHECK(req[0] == kWriteTo || req[0] == kWriteInplace);
+    const int axis = 0;
+    const NDArray &data = inputs[0];
+    const NDArray &idx = inputs[1];
+    const NDArray &out = outputs[0];
+    CHECK_EQ(axis, 0) << "Not supported yet";
+    CHECK_EQ(data.shape()[axis], idx.shape()[0]);
+    CHECK_EQ(idx.shape().ndim(), 1U);
+    Stream<gpu>* s = ctx.get_stream<gpu>();
+    cudaStream_t stream = Stream<gpu>::GetStream(s);
+    // count the number of 1s in `idx`, so that we could know the output dimension
+    size_t idx_size = idx.shape()[0];
+    int32_t valid_num = 0;
+    int32_t* prefix_sum = nullptr;
+    void* d_temp_storage = nullptr;
+    size_t temp_storage_bytes = 0;
+    // Calculate total temporary memory size
+    cub::DeviceScan::InclusiveSum(d_temp_storage,
+                                    temp_storage_bytes,
+                                    prefix_sum,
+                                    prefix_sum,
+                                    idx_size,
+                                    stream);
+    size_t buffer_size = idx_size * sizeof(int32_t);
+    temp_storage_bytes += buffer_size;
+    // Allocate memory on GPU and allocate pointer
+    Tensor<gpu, 1, char> workspace =
+        ctx.requested[0].get_space_typed<gpu, 1, char>(Shape1(temp_storage_bytes), s);
+    prefix_sum = reinterpret_cast<int32_t*>(workspace.dptr_);
+    d_temp_storage = workspace.dptr_ + buffer_size;
+    MSHADOW_TYPE_SWITCH_WITH_BOOL(idx.dtype(), IType, {
```

Comment:
You may also simplify the `MSHADOW_TYPE_SWITCH_WITH_BOOL` here. This is inside the check of `inputs[np_indexing_::kIdx].dtype() == mshadow::kBool`. Thus, IType will always be bool